### PR TITLE
CT-812 add guards to state machine event transitions

### PR DESCRIPTION
**Pull request by misaka**
Initial commit to add functionality off having per-transition guards.

### CODE REVIEW

File: spec/lib/case_state_machine_spec.rb [hunk 1]
```diff
@@ -177,15 +177,19 @@ def event(event_name)
 
     it { should transition_from(:pending_dacu_clearance)
                   .to :pending_press_office_clearance }
-    it { should require_permission(:can_escalate_to_next_approval_level?)
+    xit { should require_permission(:can_escalate_to_next_approval_level?)
```
Comment (aliuk2012) [inline spec/lib/case_state_machine_spec.rb]: xit !!

File: spec/lib/case_state_machine_spec.rb [hunk 2]
```diff
@@ -177,15 +177,19 @@ def event(event_name)
...
-    it { should transition_from(:pending_dacu_clearance).to :awaiting_dispatch}
-    it { should transition_from(:pending_press_office_clearance).to :awaiting_dispatch}
-    it { should require_permission(:can_approve_case?)
+    it { should transition_from(:pending_dacu_clearance)
+                  .to(:awaiting_dispatch)
+                  .checking_case_policy(:can_approve_case?) }
+    it { should transition_from(:pending_press_office_clearance)
+                  .to(:awaiting_dispatch)
+                  .checking_case_policy(:can_approve_case?) }
+    xit { should require_permission(:can_approve_case?)
```
Comment (aliuk2012) [inline spec/lib/case_state_machine_spec.rb]: same as above

File: spec/site_prism/page_objects/pages/cases/show_page.rb [hunk 3]
```diff
@@ -8,6 +8,7 @@ class ShowPage < SitePrism::Page
                 PageObjects::Sections::PrimaryNavigationSection, '.global-nav'
 
         element :escalation_notice, '.alert-orange'
+        elements :notices, '.notice-summary-heading'
```
Comment (aliuk2012) [inline spec/site_prism/page_objects/pages/cases/show_page.rb]: shouldn't this be `element`? i.e only expect on notice?

Comment (misaka) [inline spec/site_prism/page_objects/pages/cases/show_page.rb]: Ya, I was thrown off by the `.each` I showed you, but you're right ... it should only ever get one of those keys. :100:

**Comment by aliuk2012** [comment]
@misaka code coverage has now dropped down by 22.4%

**Comment by misaka** [comment]
Presumably the code coverage dropping is because of the error.

**Comment by aliuk2012** [comment]
Tests are passing now but code coverage still dropped 22%

**Comment by misaka** [comment]
So why does https://codeclimate.com/github/ministryofjustice/correspondence_tool_staff show 98% \*scratching-head\*

**Pull request merged**